src/main/kotlin/nl/nuts/discovery/store/entity/NutsCertificateRequest.kt

```diff
@@ -0,0 +1,159 @@
+import org.bouncycastle.openssl.PEMParser
+import org.bouncycastle.pkcs.PKCS10CertificationRequest
+import java.io.BufferedReader
+import java.io.ByteArrayInputStream
+import java.io.InputStreamReader
+import java.io.StringReader
+import java.time.LocalDateTime
+import javax.persistence.Column
+import javax.persistence.Entity
+import javax.persistence.GeneratedValue
+import javax.persistence.GenerationType
+import javax.persistence.Id
+
+
+/**
+ * Entity to store PEM encoded CSR in DB
+ */
+@Entity
+class NutsCertificateRequest {
+
+    companion object {
+        val NUTS_VENDOR_EXTENSION: ASN1ObjectIdentifier = ASN1ObjectIdentifier("1.3.6.1.4.1.54851.4").intern()
+
+        /**
+         * Create entity from PKCS10CertificationRequest, stores .encoded as bytes
+         */
+        fun fromPEM(pem: String): NutsCertificateRequest {
+            val req = NutsCertificateRequest().apply {
+                val csr = parsePEM(pem)
+
+                name = CordaX500Name.parse(csr.subject.toString()).toString() // this puts stuff in right order
+                this.pem = pem
+                submittedAt = LocalDateTime.now()
+                oid = extractOID(csr)
+            }
+
+            if (req.oid == null) {
+                throw IllegalArgumentException("Given CSR does not have a correctly formatted oid in extensions")
+            }
+
+            return req
+        }
+
+        /**
+         * Parse a pem-encoded CSR
+         */
+        fun parsePEM(pem: String): PKCS10CertificationRequest {
+            ByteArrayInputStream(pem.toByteArray(Charsets.UTF_8)).use {
+                val pemParser = PEMParser(BufferedReader(InputStreamReader(it)))
+                val parsedObj = pemParser.readObject()
+                if (parsedObj is PKCS10CertificationRequest) {
+                    return parsedObj
+                }
+
+                throw IllegalArgumentException("Couldn't parse PEM as CERTIFICATE REQUEST")
+            }
+        }
+
+        /**
+         * Find the OID in the subjectAltName.otherName extension
+         */
+        fun extractOID(csr: PKCS10CertificationRequest): String? {
+            var vendor: String? = null
+            val certAttributes = csr.attributes
+            for (attribute in certAttributes) {
+                if (attribute.attrType.equals(PKCSObjectIdentifiers.pkcs_9_at_extensionRequest)) {
+                    vendor = nutsVendorFromExtReq(attribute)
+                    if (vendor != null) {
+                        break
+                    }
+                }
+            }
+
+            return vendor
+        }
+
+        private fun nutsVendorFromExtReq(attribute: Attribute) : String? {
+            var vendor: String? = null
+
+            val extensions = Extensions.getInstance(attribute.attrValues.getObjectAt(0))
+            val gns = GeneralNames.fromExtensions(extensions, Extension.subjectAlternativeName)
+            val names = gns.names
+            for (san in names) {
```

Comment:
This function is glorious ;-)

Reply:
all thanks to the legion of the bouncy castle!